agent/lib_php_amqplib.c

```diff
@@ -0,0 +1,401 @@
+
+  /* See if we got a non-empty/non-null string for version. */
+  if (SUCCESS == result) {
+    if (nr_php_is_zval_non_empty_string(&retval)) {
+      version = Z_STRVAL(retval);
+    }
+  }
+
+  if (NRINI(vulnerability_management_package_detection_enabled)) {
+    /* Add php package to transaction */
+    nr_txn_add_php_package(NRPRG(txn), PHP_PACKAGE_NAME, version);
+  }
+
+  nr_txn_suggest_package_supportability_metric(NRPRG(txn), PHP_PACKAGE_NAME,
+                                               version);
+
+  zval_dtor(&retval);
+}
+
+static inline void nr_php_amqplib_get_host_and_port(
+    zval* amqp_connection,
+    nr_segment_message_params_t* message_params) {
+  zval* amqp_server = NULL;
+  zval* amqp_port = NULL;
+  zval* connect_constructor_params = NULL;
+
+  if (NULL == amqp_connection || NULL == message_params) {
+    return;
+  }
+
+  if (nr_php_is_zval_valid_object(amqp_connection)) {
+    connect_constructor_params
+        = nr_php_get_zval_object_property(amqp_connection, "construct_params");
+    if (nr_php_is_zval_valid_array(connect_constructor_params)) {
+      amqp_server
+          = nr_php_zend_hash_index_find(Z_ARRVAL_P(connect_constructor_params),
+                                        AMQP_CONSTRUCT_PARAMS_SERVER_INDEX);
+      if (nr_php_is_zval_non_empty_string(amqp_server)) {
+        message_params->server_address
+            = ENSURE_PERSISTENCE(Z_STRVAL_P(amqp_server));
+      }
+      amqp_port
+          = nr_php_zend_hash_index_find(Z_ARRVAL_P(connect_constructor_params),
+                                        AMQP_CONSTRUCT_PARAMS_PORT_INDEX);
+      if (nr_php_is_zval_valid_scalar(amqp_port)) {
+        message_params->server_port = Z_LVAL_P(amqp_port);
+      }
+    }
+  }
+}
+
+/*
+ * PhpAmqpLib\Channel\AMQPChannel::basic_publish
+ * Publishes a message
+ *
+ * @param AMQPMessage $msg
+ * @param string $exchange
+ * @param string $routing_key
+ * @param bool $mandatory
+ * @param bool $immediate
+ * @param int|null $ticket
+ * @throws AMQPChannelClosedException
+ * @throws AMQPConnectionClosedException
+ * @throws AMQPConnectionBlockedException
+ *
+ */
+NR_PHP_WRAPPER(nr_rabbitmq_basic_publish) {
+  zval* amqp_exchange = NULL;
+  zval* amqp_routing_key = NULL;
+  zval* amqp_connection = NULL;
+  nr_segment_t* message_segment = NULL;
+
+  nr_segment_message_params_t message_params = {
+      .library = RABBITMQ_LIBRARY_NAME,
+      .destination_type = NR_MESSAGE_DESTINATION_TYPE_EXCHANGE,
+      .message_action = NR_SPANKIND_PRODUCER,
+      .messaging_system = RABBITMQ_MESSAGING_SYSTEM,
+  };
+
+  (void)wraprec;
+
+  amqp_exchange = nr_php_get_user_func_arg(2, NR_EXECUTE_ORIG_ARGS);
+  if (nr_php_is_zval_non_empty_string(amqp_exchange)) {
+    /*
+     * In PHP 7.x, the following will create a strdup in
+     * message_params.destination_name that needs to be freed.
+     */
+    message_params.destination_name
+        = ENSURE_PERSISTENCE(Z_STRVAL_P(amqp_exchange));
+  } else {
```

Comment:
What if `ampq_exchange` is not a string at all?

Reply:
The [parameter](https://github.com/newrelic/newrelic-php-agent/pull/1009/files#diff-c7ef4a9c5f82fb59784f43ed3beebc548f983ae03675acbfa5b6b602e6d98e8dR164) list for the function call dictates that exchange is a string.

Additionally, the check [here](https://github.com/newrelic/newrelic-php-agent/pull/1009/files#diff-c7ef4a9c5f82fb59784f43ed3beebc548f983ae03675acbfa5b6b602e6d98e8dR190) guarantees it is a valid non-empty string before we set it (and strdup for php 7.x) here.

Reply:
Only create default if it is an empty string:
[ead28c2](https://github.com/newrelic/newrelic-php-agent/pull/1009/commits/ead28c21a438e64e0da8dc8dc53a1926246e73df)